tools/override-generator/main.go

```diff
@@ -0,0 +1,280 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/bazelbuild/bazel-gazelle/repo"
+	"github.com/bazelbuild/bazel-gazelle/rule"
+	"github.com/bazelbuild/buildtools/build"
+)
+
+const (
+	_name  = "override-generator"
+	_usage = "usage: This script converts `go_repository` rules to Gazelle `go_deps` overrides to assist in the migration to Bzlmod."
+)
+
+// override kinds.
+const (
+	_gazelleOverride = "go_deps.gazelle_override"
+	_archiveOverride = "go_deps.archive_override"
+	_moduleOverride  = "go_deps.module_override"
+)
+
+// attribute constants that are used multiple times.
+const (
+	_buildFileGenerationAttr = "build_file_generation"
+	_buildFileProtoModeAttr  = "build_file_proto_mode"
+	_patchArgsAttr           = "patch_args"
+	_buildDirectivesAttr = "build_directives"
+	_directivesAttr          = "directives"
+)
+
+var _defaultValues = map[string]string{
+	_buildFileGenerationAttr: "auto",
+	_buildFileProtoModeAttr:  "default",
+}
+
+var _mapAttrToOverride = map[string]string{
+	_buildDirectivesAttr: _gazelleOverride,
+	_buildFileGenerationAttr: _gazelleOverride,
+	_patchArgsAttr: _moduleOverride,
+	"patches": _moduleOverride,
```

Comment:
Should we map `patch_args` and `patches` to archive_override so we don't need to merge module_override with archive_override later?

Reply:
If there's no `urls` archive specified, we want to use the `module_override` (rather than the `archive_override`. Naming the go_deps override `archive_override` was explicitly chosen to mirror the Bzlmod `archive_override`, meaning that we are explicitly loading this from a different archive URL.

If we don't want to do this, then we would need to adjust the implementation of the overrides and documentation.